pkg/hub/addon/discovery_controller.go

```diff
@@ -140,18 +144,40 @@ func (c *addOnFeatureDiscoveryController) syncAddOn(ctx context.Context, cluster
 		return nil
 	}
 
-	// merge labels
-	modified := false
-	cluster = cluster.DeepCopy()
-	resourcemerge.MergeMap(&modified, &cluster.Labels, labels)
-
-	// no work if the cluster labels have no change
-	if !modified {
+	// no work if the cluster labels are equal
+	if equality.Semantic.DeepEqual(cluster.Labels, labels) {
 		return nil
 	}
 
-	// otherwise, update cluster
-	_, err = c.clusterClient.ClusterV1().ManagedClusters().Update(ctx, cluster, metav1.UpdateOptions{})
+	oldData, err := json.Marshal(clusterv1.ManagedCluster{
```

Comment:
maybe we could only build the patch like https://github.com/open-cluster-management-io/registration/blob/18e798299b3a78ead5075c6db4a6bfeea9d10e4c/pkg/hub/managedcluster/controller.go#L100

Reply:
and the unit tests need to be updated too

Reply:
done.